### ISSUE TITLE
[Question] How to create the target tensor for my own data?

### ISSUE THREAD

**Issue by farazkhan86** [wolny/pytorch-3dunet#4]
Hi, 
Many thanks for providing this framework for 3D CNN. Its a huge help for people like myself who are trying to teach them self deep learning with volumetric data. 

I understand the working of your code and also got it to run on the random data but I am struggling with how to create the target tensor for my 3D data. I taught myself semantic segmentation with 2D data using the CamVid dataset. And I understood the process of labeling for semantic segmentation for 2D images. I was able to create label or target tensors using a colour coded method provided for the dataset. The colour codes provided were:

```
    ("Animal", np.array([64, 128, 64], dtype=np.uint8)),
    ("Archway", np.array([192, 0, 128], dtype=np.uint8)),
    ("Bicyclist", np.array([0, 128, 192], dtype=np.uint8)),
    ("Bridge", np.array([0, 128, 64], dtype=np.uint8)),
    ("Building", np.array([128, 0, 0], dtype=np.uint8)),
    ("Car", np.array([64, 0, 128], dtype=np.uint8)),
    ("CartLuggagePram", np.array([64, 0, 192], dtype=np.uint8)),
    ("Child", np.array([192, 128, 64], dtype=np.uint8)),
```

To create a label for my network I would simply locate these RGB colours in the images and compare each "object" with its respective colour code and then give it a value in a grayscale image. So each label would be a single channel tensor. And the output of my system would have channels equal to the number of classes in the dataset (for CamVid this was 32).

No I am trying to do the same in 3D and am struggling to understand what to do. The problem is a binary one as I have to detect a single item in the input so basically I should have two output channels: Background and foreground. I have a 3D single channel input, the input having the shape: ``[BS, Channel, Z, X, Y]``. 

For this input, I have a tensor which indicates the locations of my object of interest in the input. A 3D Gaussian is built around each point to be measured. 

I hope I have explained my situation clearly.

My confusion is how to create a label tensor in 3D for segmenting my points of interest as I have done for the CamVid dataset.

Below is a 2D representation of a target that has 3D Gaussians drawn on the points of interest, the tensor is of the size ``[16, 64, 64]``

![target](https://user-images.githubusercontent.com/44167893/48837877-776b8f00-ed7e-11e8-9a19-0170ce282e20.png)

Here is the target file as a pickle dump (just in case): https://app.box.com/s/movmlt377991ptrpd0kug0d3d8s85q9l

Similar to the CamVid case as I would have 1 channel per class, here I know that I would have 2 channels, one for the background and 1 for the foreground, but how to label these 3D tensors is what I dont understand.

**Comment by wolny** [comment]
Hi @farazkhan86 thanks for showing interest in my implementation of the 3D UNet.
For binary (and multi-class) classification use the following HDF5 backed `DataLoader`, which I've added to the repo recently: https://github.com/wolny/pytorch-3dunet/blob/master/datasets/hdf5.py#L9, or the https://github.com/wolny/pytorch-3dunet/blob/master/datasets/hdf5.py#L182 if you want some basic data augmentation during training.

Since you have a single channel raw data, the raw dataset in your H5 file should be 3-dim tensor (also make sure the axis order is `DHW` if you're using the data augmentation).

Similarly your target (label) dataset should be 3-dim tensor as well, with `0` corresponding to the 1st class (e.g. background) and `1` corresponding to the 2nd class (e.g. foreground).

Then train with `NLL` loss, e.g.:
```
python train.py --checkpoint-dir ~/3dunet --in-channels 1 --out-channels 2 --layer-order crg --loss nll --learning-rate 0.0002 --interpolate       
```
In general if you have `C` classes, then the training is exactly the same, just make sure that your target (label) dataset is 3-dim tensor with values `0..C` and use the `NLL` loss.

I need to change the README to reflect that, since I use 'Binary Cross Entropy' in my example. That's just because I created this repo to solve a very specific problem where my target (label) dataset consists of `M` binary masks for the raw volume, i.e. I want to predict `M` channel binary output. In that case I had to use Binary Cross Entropy loss, and my target (label) dataset was 4-dim tensor of the size `MxDxHxW`. Hope this would help you successfully train the network.

**Comment by farazkhan86** [comment]
This is very helpful.
Many thanks for your explanation and for sharing your code on GitHub with us :)

**Comment by wolny** [comment]
No worries, I hope it will help you solve the problem you're dealing with.
BTW, I've made some changes to the to the loss functions being used during training (apart from standard `CrossEntropyLoss` and `BinaryCrossEntropyLoss` I've added `WeightedCrossEntropyLoss` and `GeneralizedDiceLoss` implemented according to https://arxiv.org/pdf/1707.03237.pdf). Please have a look in the [README](https://github.com/wolny/pytorch-3dunet#loss-functions) for more info.

**Comment by farazkhan86** [comment]
Many thanks. :)

I see that your dice loss returns a negative value

`return -1.0 * coeff
`
Is there a reason for this as in the paper here: https://arxiv.org/abs/1606.04797 tehy dont use a negative dice loss.

Thanks again

**Comment by farazkhan86** [comment]
Hi @wolny 

Reading your vastly improved recent version of the code I decided to rewrite my code to utilize your hd5 dataset and augmentation. I have a couple questions about this.

What is the exact arrangement of the hdf5 files? because I am a bit confused with the patch size option. If I have 500 3D volumes for training, I would create a hdf5 dataset and append each volume to it. Is this the structure of the hdf5 file the network is expecting? 

Then using the patch size, and stride, the network will iterate over each sample patch wise? Once a single 3D volume is covered patch by patch it will move to the next sample in the hdf5 dataset?

is my understanding correct?

Also, as regarding the randomRotate used:

```
        if m.ndim == 3:
            m = np.rot90(m, k, (1, 2))
        else:
            for c in range(m.shape[0]):
                m[c] = np.rot90(m[c], k, (1, 2))
```

I understand that this rotates while keeping the z-plane constant? If that is the case for ``dim==4`` shouldnt the command be :  ``m[c] = np.rot90(m[c], k, (2, 3))``

Based on my image sources, I need to rotate without touching the z-axis, it shouldnt be random rotation along all axis.

Many thanks

**Comment by wolny** [comment]
Hey @farazkhan86,

re the HDF5 backed dataset: while creating it you need to provide the file system path to the raw data and label data, as well as internal H5 path for the raw and label (if the internal is not provided the implementation will try to load `/raw` dataset as the raw data and `/label` dataset for the label).
E.g. having a training file in your `/home/user/datasets/train.h5`
```
/volume #raw data
/target #lable data
```
you can create the `HDF5Dataset` by `HDF5Dataset('/home/user/datasets/train.h5', patch_shape=..., stride_shape=..., raw_internal_path='/volume', label_internal_path='/target')` (bear in mind that both raw and label are in the same H5 file in this case.

If you 500 3D volumes, each in different H5 file, I'd recommend using the `torch.utils.data.ConcatDataset` so that you don't have to create this additional giant H5 with all the training data in it. E.g. use something like:
```
h5_datasets = []
for i in range(500):
  h5_datasets.append(HDF5Dataset(f'file{i}.h5', ...))

result = ConcatDataset(h5_datasets)
```

`RandomRotate90` tranformer is implemented correctly even for 4D tensors, look closely: `m` is 4D `m[c]` is 3D, so `np.rot90(m[c], k, (1, 2))` is correct.
